### ISSUE TITLE
lintr does not handle nested source()?

### ISSUE THREAD

**Issue by ptepa** [r-lib/lintr#2702]
# lintr does not handle nested source()?

$ cat lintr_bug1.R
my_global_function <- function() {
  cat("in my_global_function\n")
}

$ cat lintr_bug2.R
source("lintr_bug1.R")
my_global_function2 <- function() {
  my_global_function()
}

$ cat lintr_bug.R
source("lintr_bug2.R")
my_global_function2()

$ Rscript -e "library(lintr) ; lint_dir('.')"
lintr_bug2.R:3:3: warning: [object_usage_linter] no visible global function definition for ‘my_global_function’
  my_global_function()
  ^~~~~~~~~~~~~~~~~~

**Comment by MichaelChirico** [comment]
There's nothing about nesting -- we don't follow `source()` at all.

```r
writeLines("foo = function(x) x", tmp1<-tempfile())
writeLines(c(sprintf("source('%s')", tmp1), "bar = function() foo()"), tmp2<-tempfile())

lint(tmp2, object_usage_linter())
# /tmp/RtmpZLenmx/file81f08c22f8e2:2:18: warning: [object_usage_linter] no visible global function definition for 'foo'.
# bar = function() foo()
#                  ^~~
```

I think doing so is out of scope -- in general it's quite hard to do so well+correctly. We have some support for [{box}](https://github.com/klmr/box) and of course you can use a package instead of `source()`-ing in helpers.

**Comment by ptepa** [comment]
$ cat undefined.R

if (x) { # Why doesn't lintr report that x is undefined here?
  cat("x is defined as non-false.\n")
}

x <- 1L

$ Rscript -e "library(lintr) ; lint_dir('.')"
undefined.R:7:1: style: [trailing_blank_lines_linter] Trailing blank lines are superfluous.

^
